ext/page_log/palite/palite.cpp

```diff
@@ -1019,741 +799,1190 @@ class Storage {
+          /*
+           * Globals table to store LSN counter and known table IDs.
+           *  - LSN counter id = 0
+           *  - Table IDs id = 1
+           */
+          R"(CREATE TABLE IF NOT EXISTS globals (
+                 id INTEGER CHECK (id = 0 OR id = 1),
+                 val INTEGER NOT NULL,
+             PRIMARY KEY (id, val));)",
 
           /* Init LSN counter, if the table is empty; do nothing otherwise. */
-          R"(INSERT OR IGNORE INTO lsn(id, lsn) VALUES (0, 0);)"
+          R"(INSERT INTO globals (id, val)
+             SELECT 0, 0
+             WHERE NOT EXISTS (SELECT 1 FROM globals WHERE id = 0);)"};
 
+        ~Globals() = default;
+        Globals(Config &cfg, const std::filesystem::path &home)
+            : Table<Globals>(cfg, home / std::format("{}.db", Globals::prefix))
+        {
+        }
+
+        Globals(const Globals &) = delete;
+        Globals &operator=(const Globals &) = delete;
+
+        auto
+        conn_config() -> std::ranges::view auto
+        {
+            /* No special configuration is required for the 'globals' table. */
+            return std::views::empty<std::string_view>;
+        }
+
+        uint64_t
+        fetch_lsn(Connection &conn, Statement stmt_type)
+        {
+            Connection::StatementPtr stmt = conn.db_statement(stmt_type);
+            SQ_CHECK(sqlite3_step, stmt.get());
+            return sqlite3_column_int64(stmt.get(), 0);
+        }
+
+        uint64_t
+        make_next_lsn(Connection &conn)
+        {
+            return fetch_lsn(conn, Statement::MAKE_NEXT_LSN);
+        }
+
+        uint64_t
+        get_last_lsn(Connection &conn)
+        {
+            return fetch_lsn(conn, Statement::GET_LAST_LSN);
+        }
+
+        void
+        add_table_id(Connection &conn, uint64_t table_id)
+        {
+            Connection::StatementPtr stmt = conn.db_statement(Statement::ADD_TABLE_ID);
+            SQ_CHECK(sqlite3_bind_int64, stmt.get(), 1, static_cast<sqlite3_int64>(table_id));
+            SQ_CHECK(sqlite3_step, stmt.get());
+        }
+
+        std::vector<uint64_t>
+        get_table_ids(Connection &conn)
+        {
+            Connection::StatementPtr stmt = conn.db_statement(Statement::GET_TABLE_IDS);
+            std::vector<uint64_t> table_ids;
+            while (SQ_CHECK(sqlite3_step, stmt.get()) == SQLITE_ROW) {
+                table_ids.push_back(static_cast<uint64_t>(sqlite3_column_int64(stmt.get(), 0)));
+            }
+            return table_ids;
+        }
+    };
+
+    struct Checkpoints : public Table<Checkpoints> {
+        constexpr static std::string_view prefix = "checkpoints";
+
+        enum Statement { PUT_CHECKPOINT, GET_CHECKPOINT, DELETE_CHECKPOINTS };
+
+        constexpr static std::string_view sql_statements[] = {
+          /*
+           * PUT_CHECKPOINT: Insert a new checkpoint.
+           */
+          R"(INSERT INTO checkpoints (lsn, timestamp, checkpoint_metadata)
+             VALUES (?, ?, ?);)",
```

Comment:
I personally prefer when we have statements assigned directly, like:
```
sql_statements[PUT_CHECKPOINT] = R"...";
```

Since it reduces the possibility to make an ordering error when adding new ones.